src/core/thread/network_data_publisher.cpp

```diff
@@ -0,0 +1,1016 @@
+
+    mType   = kTypeExternalRoute;
+    mPrefix = aConfig.GetPrefix();
+    mStable = aConfig.mStable;
+    mFlags  = aConfig.ConvertToTlvFlags();
+
+    SetState(kToAdd);
+
+    Process();
+}
+
+void Publisher::PrefixEntry::Unpublish(void)
+{
+    bool registerWithLeader = false;
+
+    otLogInfoNetData("Publisher: Unpublishing %s", mPrefix.ToString().AsCString());
+
+    registerWithLeader = Remove();
+
+    if (registerWithLeader)
+    {
+        Get<Notifier>().HandleServerDataUpdated();
+    }
+}
+
+void Publisher::PrefixEntry::HandleNotifierEvents(Events aEvents)
+{
+    if (aEvents.ContainsAny(kEventThreadNetdataChanged | kEventThreadRoleChanged))
+    {
+        Process();
+    }
+}
+
+bool Publisher::PrefixEntry::Add(void)
+{
+    // Adds the prefix entry to the network data.
+
+    bool registerWithLeader = false;
+
+    union
+    {
+        OnMeshPrefixConfig  mOnMeshPrefix;
+        ExternalRouteConfig mExternalRoute;
+    } config;
+
+    switch (mType)
+    {
+    case kTypeOnMeshPrefix:
+        config.mOnMeshPrefix.mPrefix = mPrefix;
+        config.mOnMeshPrefix.mStable = mStable;
+        config.mOnMeshPrefix.SetFromTlvFlags(mFlags);
+        SuccessOrExit(Get<Local>().AddOnMeshPrefix(config.mOnMeshPrefix));
+        break;
+
+    case kTypeExternalRoute:
+        config.mExternalRoute.mPrefix = mPrefix;
+        config.mExternalRoute.mStable = mStable;
+        config.mExternalRoute.SetFromTlvFlags(static_cast<uint8_t>(mFlags));
+        SuccessOrExit(Get<Local>().AddHasRoutePrefix(config.mExternalRoute));
+        break;
+    }
+
+    registerWithLeader = true;
+    SetState(kAdded);
+
+exit:
+    return registerWithLeader;
+}
+
+bool Publisher::PrefixEntry::Remove(State aNextState)
+{
+    // Remove the prefix entry to the network data.
+
+    bool registerWithLeader = false;
+
+    VerifyOrExit((GetState() == kAdded) || (GetState() == kRemoving));
+
+    switch (mType)
+    {
+    case kTypeOnMeshPrefix:
+        IgnoreError(Get<Local>().RemoveOnMeshPrefix(mPrefix));
+        break;
+
+    case kTypeExternalRoute:
+        IgnoreError(Get<Local>().RemoveHasRoutePrefix(mPrefix));
+        break;
+    }
+
+    registerWithLeader = true;
+
+exit:
+    SetState(aNextState);
+    return registerWithLeader;
+}
+
+void Publisher::PrefixEntry::Process(void)
+{
+    // This method checks the entries currently present in Network Data
+    // based on which it then decides whether or not take action
+    // (add/remove or keep monitoring).
+
+    uint8_t numEntries          = 0;
+    uint8_t numPreferredEntries = 0;
+    uint8_t desiredNumEntries   = 0;
+
+    // Do not make any changes if device is not attached, and wait
+    // for role change event.
+    VerifyOrExit(Get<Mle::Mle>().IsAttached());
+
+    VerifyOrExit(GetState() != kNoEntry);
+
+    switch (mType)
+    {
+    case kTypeOnMeshPrefix:
+        CountOnMeshPrefixEntries(numEntries, numPreferredEntries);
```

Comment:
What if this prefix entry is already added to the `Network Data`  but the state is not `kAdded` ? This may forbid this Prefix entry to change to its correct state `kAdded`. I think such case is possible when the device received a network data after reset.

Reply:
I realized that `Publisher` only need to be consistent with Local network data, so no more issue.